### ISSUE TITLE
Change in-app notification height

### ISSUE THREAD

**Issue by amivaleo** [junrrein/pdfslicer#134]
Hi!

Whenever you save a pdf, this notification appears on top:

![Schermata da 2019-10-01 11-19-38](https://user-images.githubusercontent.com/24316547/65950351-2ddbde00-e43e-11e9-8191-248a113be1db.png)

Can you set its size so to be consistent with the same widget used in other applications?
For example, in geary:

![Schermata da 2019-10-01 11-19-12](https://user-images.githubusercontent.com/24316547/65950409-46e48f00-e43e-11e9-8b14-ce2b9b1ea961.png)

Side-note: I argue about the tick icon too. 😁

**Comment by junrrein** [comment]
I wouldn't mind being consistent with the other apps :+1: 

> Side-note: I argue about the tick icon too. :grin:

Do you argue that it should be a close-button instead? If it's that, then I agree.

**Comment by amivaleo** [comment]
> I wouldn't mind being consistent with the other apps

:(

> Do you argue that it should be a close-button instead? If it's that, then I agree.

No...

You seem upset. Maybe... It's just an impression. Never mind.

**Comment by junrrein** [comment]
> > I wouldn't mind being consistent with the other apps
> 
> :(

I was pretty sure that "I wouldn't mind if.." means "It would be nice if...". I'm not so sure anymore :sweat_smile: 

In this case, I meant that it would be nice to be consistent with the other apps :+1: 

> > Do you argue that it should be a close-button instead? If it's that, then I agree.
> 
> No...

What do you have in mind?

> You seem upset. Maybe... It's just an impression. Never mind.

Not at all. It's just meanings lost in translation.

**Comment by amivaleo** [comment]
:3

My fault. I translated "I wouldn't mind" like "I don't care".

Sorry. 😶

So, I guess that the "standard" way in which that notification is used is: some_text [X]. Where [X] is a close button that closes the notification.

An example for this: nautilus when deleting some file or folder.

An extra button can be placed there. If I remeber correctly, no more than one single extra button.

Uhm... Yes, it is like that: https://developer.gnome.org/hig/stable/in-app-notifications.html.en

In order to follow that standard, the [X] button should be added and the height should be adjusted (maybe it gets adjusted automatically when that [X] is added...?).

About the tick-icon: I think it might be create some confusion with the [X]. Personally, I like it. I like apps being a bit "funny". The truth is that they are not, ever. :| They're emotionless. I assume there's some strong reason behind that, which I don't know.

So... The label should be only "Saved." (no exclamation mark :\)

**Comment by junrrein** [comment]
I agree about all your comments, including the ones about apps being emotionless generally. I guess the intention is to convey professionalism? I can understand it in core system components like the desktop environment, but in hobby projects like mine it's less important.

Either way, that's the way we will go forward: "Saved [X]". No tick, no exclamation mark.

Thanks for the suggestion!

**Comment by amivaleo** [comment]
> in hobby projects like mine it's less important

Forgive me if I'm too pushy. You're the man with powers here and Slicer is your kiddo, so yours is every last word.

I like to have consistency among several apps, because I strongly believe that it is a fundamental ingredient for getting more people into the linux world (through gnome, in this case). But you can say "no" at any time and follow your time. I'll just stop bothering in that case!

Apologies, Julian. :)

**Comment by junrrein** [comment]
> You're the man with powers here and Slicer is your kiddo

I'd rather not see things that way. In this case, I do agree with your assesment, so that's why I'm going through with it. Keep those suggestions coming :)

**Comment by amivaleo** [comment]
> Keep those suggestions coming :)

Ok, sure! It's a pleasure for me! 💪

**Comment by junrrein** [comment]
Hi there,

I'm working on this, but I hit a little bit of trouble.

It turns out that, as you said, letting the notification height be adjusted automatically by button size was the correct idea. When a notification has at least one button, the theme sizes the notification nicely.

The problem appears when there isn't a buton in the notification, as is the case with the "Saving..." notification. In that case, the theme doesn't care and just shrinks the notification size to its contents.

Some pictures:

![Captura de pantalla de 2019-12-28 11-54-41](https://user-images.githubusercontent.com/962049/71545349-e4c9f600-2968-11ea-82a6-84d3f9721081.png)
![Captura de pantalla de 2019-12-28 11-49-09](https://user-images.githubusercontent.com/962049/71545353-f01d2180-2968-11ea-9cff-88bb74b9565a.png)

A workaround is to set the height manually, but that only works for the specific theme I'm adjusting it for:

![Captura de pantalla de 2019-12-28 11-49-06](https://user-images.githubusercontent.com/962049/71545366-0dea8680-2969-11ea-8afd-772fd8342a55.png)
![Captura de pantalla de 2019-12-28 11-50-20](https://user-images.githubusercontent.com/962049/71545376-1c38a280-2969-11ea-88bb-75b42264af28.png)
![Captura de pantalla de 2019-12-28 11-50-23](https://user-images.githubusercontent.com/962049/71545378-1f339300-2969-11ea-9888-c18a9cf44599.png)

So there are three options I'm seeing:

* Let the "Saving..." notification look bad on all themes.
* Adjust the "Saving..." notification for Adwaita, and let it look bad on all other themes.
* Explore replacing that notification with something else.

What do you think?

**Comment by amivaleo** [comment]
> Adjust the "Saving..." notification for Adwaita, and let it look bad on all other themes.

Nope! Never. I'm 100% against hardcoded theme adjustments. Not only because they compromise the look of any other theme, but also because the very same default theme (adwaita) could be updated and hence your app might look bad.

From my point of view, coding any css stuff is close to hacking your own app.

> Let the "Saving..." notification look bad on all themes. / Explore replacing that notification with something else.

I should check what other apps do to have a some fresh example in my head. Based on my bad memory, apps usually show that spinner icone somewhere in the GUI. You could either do the same (what about showing that spinner to the right of the "Save As" button, only when slicer is saving?) or keep the in-app notification look "bad".

I checked the HIG again, and I can clearly read that the in-app notification "can ALSO include" a button, but it's not mandatory.

BUT

Yet from the HIG: "In-app notifications are not a good solution for communicating ongoing states.".

I assume that "saving" is a "state" of your app. Thus I suggest to follow the idea of showing the spinner somewhere. I suggest close to the Save As button because user focus should already be there. Another nice idea could be to place it close to the headerbar title, maybe at the beginning (without moving the entire title! It has to remain aligned in the center), so that, when saving is complete, both the spinner and the * in the title disappear together.

**Comment by junrrein** [comment]
> > Adjust the "Saving..." notification for Adwaita, and let it look bad on all other themes.
> 
> Nope! Never. I'm 100% against hardcoded theme adjustments. Not only because they compromise the look of any other theme, but also because the very same default theme (adwaita) could be updated and hence your app might look bad.

Huh, I haven't thought about the last part. And yes, compromising the look of other themes will make for unhappy users.

> I checked the HIG again, and I can clearly read that the in-app notification "can ALSO include" a button, but it's not mandatory.

Yeah, it's not mandatory, but it seems that themes only care about notifications that _do_ have a button.

> BUT
> 
> Yet from the HIG: "In-app notifications are not a good solution for communicating ongoing states.".
>
> I assume that "saving" is a "state" of your app. Thus I suggest to follow the idea of showing the spinner somewhere.

I agree 100%, but I don't know of a good place to have that spinner.

> I suggest close to the Save As button because user focus should already be there.

This one could be a solution.

> Another nice idea could be to place it close to the headerbar title, maybe at the beginning (without moving the entire title! It has to remain aligned in the center), so that, when saving is complete, both the spinner and the * in the title disappear together.

This could also be a solution, but I don't like this one as much. What about showing a progress bar at the bottom of the headerbar title? It is technically possible, since I can tap into the saving process to get the progress, I'm just not doing it at the moment.

> I should check what other apps do to have a some fresh example in my head.

I can't think of many apps which have a saving process that takes a long time. The only examples I have so far with are:

* GIMP: It blocks the save file dialog (and the app) while saving.
* Evince: It saves the file in the background without telling the user about progress or success (tested this with a humongous file, which shouldn't be saved instantly).

So these are two more alternatives we have:

* Block the app while saving.
* Don't show saving progress.

For now, I'll be commiting the notification fix that "looks bad on all themes", while we talk about what we do next.

**Comment by amivaleo** [comment]
> What about showing a progress bar at the bottom of the headerbar title?

epiphany used to do that. I think its maintainers changed it a bit so that the progress bar is now shown at the bottom of the url bar so to make it more visible.

I must have seen the same think somewhere else. Youtube maybe? I should check (my internet connection is awful in this place...). 🤔️

So, I'm a bit concerned about the visibility of such bar. I'd prefer a percentage next to a spinner. Maybe both?

> I can't think of many apps which have a saving process that takes a long time.

That's why I can't really test any app right now. Also because my laptops has a NVMe memory as hardisk, which is pretty fast. I have to try some 16k resolution picture on gimp. o.O

And yet, gimp shows a progress bar on the bottom when you overwrite the file. I wouldn't care too much about gimp, since it doesn't follow gnome early GUI style.
Freezing the save dialog might be a solution... That I don't like, since it gives me the idea that the app has encountered some error. A spinner gives more the idea that something is working on the background maybe. 🤔️

Geary uses a spinner to notify the user that it is syncing the emails. Although the spinner appears in a very awkward position...

I should check gedit. I have the idea that it shows a spinner next to the Save button. Not sure.

> For now, I'll be commiting the notification fix that "looks bad on all themes", while we talk about what we do next.

💪️💪️💪️

**Comment by junrrein** [comment]
> epiphany used to do that. I think its maintainers changed it a bit so that the progress bar is now shown at the bottom of the url bar so to make it more visible.
> (...)
> So, I'm a bit concerned about the visibility of such bar. I'd prefer a percentage next to a spinner. Maybe both?

I'm concerned both for the visibility of the bar, and that its meaning is understood. I'm not sure a lone bar or percentage is enough for an operation that rarely shows up. While you save all the time when using Slicer, it's not common to see the saving progress since most files you work on aren't big. On Epiphany, you see the progress bar all the time, so the meaning can't be misunderstood.

> That's why I can't really test any app right now. Also because my laptops has a NVMe memory as hardisk, which is pretty fast. I have to try some 16k resolution picture on gimp. o.O

I have a little bit of envy lol.

> And yet, gimp shows a progress bar on the bottom when you overwrite the file. I wouldn't care too much about gimp, since it doesn't follow gnome early GUI style.

Yeah, I wouldn't care too much about what Gimp does, but it's valid as an example since we don't have so much to work with here.

About what Gimp does: it's not that bad of an idea, but let's keep looking

> Freezing the save dialog might be a solution... That I don't like, since it gives me the idea that the app has encountered some error. A spinner gives more the idea that something is working on the background maybe. :thinking:

I agree.

> Geary uses a spinner to notify the user that it is syncing the emails. Although the spinner appears in a very awkward position...

I don't know how to trigger that spinner in Geary, but I'll take your word for it.

> I should check gedit. I have the idea that it shows a spinner next to the Save button. Not sure.

This is what gedit does when loading and saving big files:

![Captura de pantalla de 2019-12-28 14-52-36](https://user-images.githubusercontent.com/962049/71547771-31222f80-2983-11ea-8f7b-71ca064d94b1.png)
![Captura de pantalla de 2019-12-28 14-52-53](https://user-images.githubusercontent.com/962049/71547774-33848980-2983-11ea-89b2-046ab83116ff.png)

**Comment by amivaleo** [comment]
Uh! I like the second picture of gedit. That is definetely good. I suggest that as the solution: it is close to a in-app notification, you can put long strings there, there's space for a progress bar, a spinner, a percentage or whatever else you might want to put there.

But do not show that bar for every file. As you said, some files are quite light/small, so that thing would appear only for a fraction of a second -> very bad thing to see, also because it suggests the user that there's some bug in your application that doesn't allow him to read something extremely important.
Remember to code some "waiting time" or some sort of "guess" for the time needed before that banner pops up, if you go for this solution.

> I don't know how to trigger that spinner in Geary, but I'll take your word for it.

You need to have a bad connection. :| Or, try to sync all the emails from some old email accounts having tens of thousands emails. You'll see a spinner in the bottom left corner of the gui, where you have a otherwise-fully-empty bar.

> I have a little bit of envy lol.

Dell XPS 13. Pricey, but it worths every penny. I needed a good & compact laptop for my phd.

**Comment by junrrein** [comment]
> Uh! I like the second picture of gedit. That is definetely good. I suggest that as the solution: it is close to a in-app notification, you can put long strings there, there's space for a progress bar, a spinner, a percentage or whatever else you might want to put there.
> 
> But do not show that bar for every file. As you said, some files are quite light/small, so that thing would appear only for a fraction of a second -> very bad thing to see, also because it suggests the user that there's some bug in your application that doesn't allow him to read something extremely important.
> Remember to code some "waiting time" or some sort of "guess" for the time needed before that banner pops up, if you go for this solution.

Thanks for the suggestions! I'll end up doing it this way then. Yeah, there definitely needs to be some waiting time before the banner pops up, for the reasons you brought up.

I probably won't be doing it soon, since the current way works, but the new way would be a nice to-have.

Finally, I'm closing #141 since everything about a progress bar for saving operations has been discussed here.

**Comment by amivaleo** [comment]
I feel like I only supported your decision at the end. 😂️

Thank you for your app, Julian. ;)
